File: modules/libs/core/adapter/webui/assets.go
package webui

import (
	"errors"
	"fmt"
	"net/http"
	"net/url"
	"strconv"
	"strings"
	"time"
)

// A page of a document the vault holds is bytes, and bytes are what this route
// answers:
//
//	GET /assets/<id>/pages/<n>?wide=W&size=S&mtime=T   one page, drawn to that width
//
// The id is the file's path in the vault, escaped. A file has no other name the
// window holds.
//
// Nothing else of a file is here. What a file is, where a run of its text sits,
// and what a model made from it are questions of the schema; a page is a
// picture, and a picture is what an `img` loads.
const assetsRoute = "/assets/"

// The one facet an asset offers.
const pagesFacet = "pages"

// An address is one question about one asset: which file, which facet, and what
// the facet was named with.
type address struct {
	path  string
	facet string
	at    string
}

// addressed takes an asset's address apart.
//
// The escaped path is what is read: Go decodes before a handler is reached, and
// a decoded separator runs the id and the facet after it together.
func addressed(r *http.Request) (address, bool) {
	rest := strings.TrimPrefix(r.URL.EscapedPath(), assetsRoute)
	if rest == "" || rest == r.URL.EscapedPath() {
		return address{}, false
	}
	parts := strings.Split(rest, "/")
	path, err := url.PathUnescape(parts[0])
	if err != nil || path == "" {
		return address{}, false
	}
	held := address{path: path}
	if len(parts) > 1 {
		held.facet = parts[1]
	}
	if len(parts) > 2 {
		held.at = parts[2]
	}
	if len(parts) > 3 {
		return address{}, false
	}
	return held, true
}

// Asset answers with the bytes of one page of one file of the vault.
func (a *API) Asset(w http.ResponseWriter, r *http.Request) {
	at, ok := addressed(r)
	if !ok {
		http.Error(w, "not an asset", http.StatusBadRequest)
		return
	}
	if at.facet != pagesFacet {
		http.Error(w, "an asset answers with a page and nothing else", http.StatusNotFound)
		return
	}
	a.Page(w, r, at.path, at.at)
}

// assetOf is where a file of the vault is drawn from, and pageOf one page of
// it. A path is escaped whole, so a file in a folder is one segment and what
// hangs off it is the next.
func assetOf(path string) string { return assetsRoute + url.PathEscape(path) }

func pageOf(path string, at, wide int, print fingerprint) string {
	return fmt.Sprintf("%s/%s/%d?wide=%d&%s", assetOf(path), pagesFacet, at, wide, printing(print))
}

// An address that names which bytes it is about answers those bytes or nothing,
// so what it answers with may be kept for as long as anything keeps anything.
const immutable = "public, max-age=31536000, immutable"

// errChanged is a file that is no longer the one the address was given out for.
// The address is gone: the caller asks the vault again and is given another.
var errChanged = errors.New("the file changed since this address was given out")

// printing is a fingerprint as an address carries it, and printed is it read
// back. The path is a segment of the address already, so what is written here
// is the rest of what says which bytes the file is.
func printing(print fingerprint) string {
	return fmt.Sprintf("size=%d&mtime=%d", print.size, print.mtime)
}

func printed(query url.Values) (fingerprint, error) {
	size, err := strconv.ParseInt(query.Get("size"), 10, 64)
	if err != nil {
		return fingerprint{}, fmt.Errorf("size: %q is not a size", query.Get("size"))
	}
	mtime, err := strconv.ParseInt(query.Get("mtime"), 10, 64)
	if err != nil {
		return fingerprint{}, fmt.Errorf("mtime: %q is not a time", query.Get("mtime"))
	}
	return fingerprint{size: size, mtime: mtime}, nil
}

// stamp and instant are a modification time as this adapter carries it — an
// address and the schema both say nanoseconds since the epoch — and as the core
// holds one. Zero is a file nothing was said about, not the epoch.

func stamp(t time.Time) int64 {
	if t.IsZero() {
		return 0
	}
	return t.UnixNano()
}

func instant(nanos int64) time.Time {
	if nanos == 0 {
		return time.Time{}
	}
	return time.Unix(0, nanos)
}

File: modules/apps/desktop/editor/src/assets.ts
/**
 * What a file of the vault is, for whatever opens it: a document's pages, a
 * recording's duration and where its sound is played from, and its
 * transcript.
 *
 * The application serves both over addresses of its own, because the window is
 * drawn from a scheme a browser loads neither pictures nor sound through.
 */
import { Code, createClient } from '@connectrpc/connect'
import type { ConnectError } from '@connectrpc/connect'
import { ArtifactService, AssetService } from '@numen/protocol'
import type {
  Cue as CueMessage,
  HighlightedPage as HighlightedPageMessage,
  ReadTextResponse,
} from '@numen/protocol'
import { transport } from '@numen/wire'
import { fingerprint, stamp } from './answers'
import type { Documents, HighlightedPage } from './document/open'
import type { Cue, Recordings } from './media/transcript'

/** What the files of the vault are, for whatever opens one. */
const assets = createClient(AssetService, transport)

/** What a model has made from the files of the vault. */
const artifacts = createClient(ArtifactService, transport)

/**
 * The documents the vault holds, over the addresses the application serves the
 * window at. A page is a picture at an address of its own, drawn to the width
 * it is asked for in device pixels.
 */
export const documents: Documents = {
  shape: async (path) => {
    const answer = await waiting(() => assets.getDocument({ path }))
    return {
      pages: answer.pages.map((one) => ({ width: one.width, height: one.height })),
      at: stamp(answer.fingerprint) ?? '',
    }
  },
  page: (path, at, wide, seen) =>
    `${asset(path)}/pages/${at}?wide=${wide}&${named(seen)}`,
  highlights: async (path, stretches) => {
    const answer = await waiting(() => assets.listHighlights({ path, at: [...stretches] }))
    return stretches.map((_, i) => answer.runs[i]?.pages.map(highlighted) ?? [])
  },
}

/** One page a highlight falls on, as the window carries it. */
const highlighted = (one: HighlightedPageMessage): HighlightedPage => ({
  page: one.index,
  rects: one.rects.map((box) => ({
    minX: box.minX,
    minY: box.minY,
    maxX: box.maxX,
    maxY: box.maxY,
  })),
})

/**
 * The recordings the vault holds, over the same addresses. The player is given
 * an address of its own: the window is drawn from a scheme a browser does not
 * load sound through, and the application answers where it does.
 */
export const recordings: Recordings = {
  listened: async (path) => {
    const answer = await waiting(() => assets.getRecording({ path }))
    return {
      duration: answer.durationMs,
      mediaUrl: answer.mediaUrl,
      mediaType: answer.mediaType,
      url: answer.url,
    }
  },
  cues: async (path) => {
    const answer = await waiting(() => artifacts.readText({ path }))
    return { ...said(answer), editable: answer.editable }
  },
  writes: async (path, cues) => {
    await waiting(() => artifacts.writeTranscript({ path, cues: [...cues] }))
  },
  plays: async (path, stretch) => {
    const answer = await waiting(() => artifacts.readText({ path, at: stretch }))
    return said(answer).cues[0]?.from ?? null
  },
}

/** One stretch of speech, kept as the plain value the window carries it as. */
const spoken = (one: CueMessage): Cue => ({ text: one.text, from: one.from, to: one.to })

/**
 * The text of a file as it came: words against the clock, or prose nothing
 * timed. A file nothing has been read or transcribed for came in neither.
 */
const said = (answer: ReadTextResponse): { cues: readonly Cue[]; prose: string } => {
  switch (answer.text.case) {
    case 'spoken':
      return { cues: answer.text.value.cues.map(spoken), prose: '' }
    case 'prose':
      return { cues: [], prose: answer.text.value }
    default:
      return { cues: [], prose: '' }
  }
}

/**
 * Where a file of the vault is asked about. The path is written out whole, so a
 * file in a folder is one part of the address and the facet asked of it is the
 * next.
 */
const asset = (path: string): string => `/assets/${encodeURIComponent(path)}`

/**
 * Which bytes an address is about, as the address writes them. The path is a
 * part of the address already, so what is written here is the rest of what says
 * which file it is.
 */
const named = (seen: string): string => {
  const at = fingerprint(seen)
  return `size=${at.size}&mtime=${at.mtime}`
}

/** How often a document that is busy is waited out before it is a refusal. */
const PATIENCE = 3

/** How long the window waits before asking a busy document again. */
const AGAIN = 1000

/**
 * What the application answered. A document held by whoever is drawing from it
 * says so, and is asked again after a wait.
 */
const waiting = async <T>(ask: () => Promise<T>): Promise<T> => {
  for (let asked = 0; ; asked++) {
    try {
      return await ask()
    } catch (error) {
      if (Code.Unavailable !== (error as ConnectError).code || asked >= PATIENCE) throw error
      await sleep(AGAIN)
    }
  }
}

const sleep = (ms: number) => new Promise((wake) => setTimeout(wake, ms))
